### PR TITLE
ISSUE-119 PGP: EncryptedProjectionCleanupListener

### PR DESCRIPTION
**Pull request by vttranlina**
Ref: https://github.com/linagora/tmail-backend/issues/119

### CODE REVIEW

File: tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/event/EncryptedProjectionCleanupListener.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        return event instanceof Expunged;
+    }
+
+    @Override
+    public Publisher<Void> reactiveEvent(Event event) {
+        if (event instanceof Expunged) {
+            Expunged expunged = (Expunged) event;
+
+            return handleMessageDeletion(expunged);
+        }
```
Comment (chibenwa) [inline tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/event/EncryptedProjectionCleanupListener.java:47]: We need to handle mailbox deletion too

File: tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/event/EncryptedProjectionCleanupListener.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        return Mono.empty();
+    }
+
+    private Mono<Void> handleMessageDeletion(Expunged expunged) {
+        return Flux.fromIterable(expunged.getExpunged()
+            .values())
+            .map(MessageMetaData::getMessageId)
+            .flatMap(messageId -> Mono.from(encryptedEmailContentStore.delete(messageId)))
```
Comment (chibenwa) [inline tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/event/EncryptedProjectionCleanupListener.java:55]: We need to delete the underlying data only if the message is no longer accessible.

Overwize you might destroy it upon a move.

Can you add a test showing that move do not delete the GPG representation of the message?

File: tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/event/EncryptedProjectionCleanupListener.java (hunk 1 above)
Comment (vttranlina) [reply tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/event/EncryptedProjectionCleanupListener.java:47]: How can we get messageId list when mailbox deletion? 
CassandraMessageIdDAO?

Comment (chibenwa) [reply tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/event/EncryptedProjectionCleanupListener.java:47]: :+1:

Comment (chibenwa) [reply tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/event/EncryptedProjectionCleanupListener.java:47]: (We don't have much choice but this makes the listener specific to Cassandra. I would be happy that we abstract the use of the DAO behind an interface to be honnest...)

Comment (vttranlina) [reply tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/event/EncryptedProjectionCleanupListener.java:47]: > we abstract the use of the DAO behind an interface

I don't really understand what you mean. Can you explain in detail?

**Comment by vttranlina** [comment]
When mailbox deletion, I used `CassandraMessageIdDAO` to retrieve the messageId list. 
But I don't sure about this way, Can the situation happen that messageId has been removed in CassandraMessageIdDAO before the `EncryptedProjectionCleanupListener` handler?

**Comment by chibenwa** [comment]
> But I don't sure about this way, Can the situation happen that messageId has been removed in CassandraMessageIdDAO before the EncryptedProjectionCleanupListener handler?

You are right, this looks shitty. If the DeleteMessageListener is executed before this one then we are screwed....

What solutions can we think of here?

**Comment by vttranlina** [comment]
> You are right, this looks shitty. If the DeleteMessageListener is executed before this one then we are screwed....
> 
> What solutions can we think of here?

If we are implementing `PreDeletionHook`, Need we keep a `ReactiveGroupEventListener` implementation? 
I think NO.

**Comment by chibenwa** [comment]
> If we are implementing PreDeletionHook, Need we keep a ReactiveGroupEventListener implementation?
>I think NO.

Yes we would not need it anymore ;-)

**Comment by vttranlina** [comment]
I will rename the commit message when squash. 
`ISSUE-119 PGP: EncryptedProjectionCleanupListener` -> `ISSUE-119 PGP: DeleteEncryptedProjectionHook`

File: tmail-backend/mailbox/encrypted/cassandra/pom.xml [hunk 3]
```diff
@@ -43,6 +43,31 @@
...
+            <type>test-jar</type>
+        </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>blob-cassandra</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>blob-memory</artifactId>
```
Comment (chibenwa) [inline tmail-backend/mailbox/encrypted/cassandra/pom.xml:63]: scope test

File: tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/DeleteEncryptedProjectionHook.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    public DeleteEncryptedProjectionHook(EncryptedEmailContentStore encryptedEmailContentStore) {
+        this.encryptedEmailContentStore = encryptedEmailContentStore;
+    }
+
+    @Override
+    public Publisher<Void> notifyDelete(DeleteOperation deleteOperation) {
+        Preconditions.checkNotNull(deleteOperation);
+        return Flux.fromIterable(deleteOperation.getDeletionMetadataList())
+            .map(metadataWithMailboxId -> metadataWithMailboxId.getMessageMetaData().getMessageId())
+            .flatMap(messageId -> Mono.from(encryptedEmailContentStore.delete(messageId)))
```
Comment (chibenwa) [inline tmail-backend/mailbox/encrypted/cassandra/src/main/java/com/linagora/tmail/encrypted/cassandra/DeleteEncryptedProjectionHook.java:27]: Please check if the metadata is still accessible before deleting it...

EG:

```
messageManagerA.append(xxx) // message is in mailboxA
mailboxManager.copy(A, B, ...) // message is now in mailbox A and B
messageManagerB.delete(xxx) // Message is now in mailbox A only and we should not remove it's content!
```

**Review by Arsnael — COMMENTED** [review]
Read it, nothing to add

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Can you squash your fixups? :)

**Comment by vttranlina** [comment]
Squash

**Pull request merged**